client/my-sites/theme/main.jsx

```diff
@@ -188,8 +198,12 @@ class ThemeSheet extends Component {
 
 	isLoading = () => {
 		const { isLoading, isThemeActivationSyncStarted } = this.props;
-		const { isAtomicTransferCompleted } = this.state;
-		return isLoading || ( isThemeActivationSyncStarted && ! isAtomicTransferCompleted );
+		const { isAtomicTransferCompleted, ssrLoadingTrigger } = this.state;
+		return (
+			isLoading ||
+			( isThemeActivationSyncStarted && ! isAtomicTransferCompleted ) ||
+			ssrLoadingTrigger
```

Comment:
The code that computes `isLoading` from the "is loading cart" and "is products list empty" values was added in #69616 by @valterlorran. I'm wondering why are we interested in these values, because it seems that the `ThemeSheet` component is not looking at any products list data or cart data -- why does it care about them being loaded?

Reply:
> On the logged-in version, there are changes in the CTA and the banner

Thanks, I think I understand what is happening now. On your screen capture, the label changes from "Upgrade to subscribe" to "Subscribe to activate". That change happens probably when the `/sites/:site/features` request (initiated from top-level layout component) finishes fetching, because only after that we know that the site is eligible for external themes. The `disabled` state doesn't watch for this, however. It's entirely possible that the wrong "Upgrade to subscribe" button is rendered as active.

The products list is not needed for rendering the Theme page, it's needed by the `onButtonClick` hander, which executes the `addExternalManagedThemeToCart` function. That function assumes that the products list is loaded. It looks at the products list to figure the right products (plan and theme) to add to the cart.

It would be better if this `addExternalManagedThemeToCart` function itself did the products list check. It's an async thunk, it only needs a few lines like this to be added:
```js
const hasProductsList = ( state ) => Object.keys( getProductsList( state ) ).length > 0;
if ( ! hasProductsList( state ) ) {
  await dispatch( requestProductsList( { type: 'all' } ) );
}
// should have valid products list now. Important to get fresh state from `getState`.
const products = getProductsList( getState() );
```
Having this, you can completely remove `QueryProductsList` and `getProductsList` from `theme/main.jsx`. After clicking the button, the handler itself will wait for products to load.

The `isLoadingCart` selector is there to disable the button after it was clicked. The `isLoadingCart` state is set to `true` by the `addExternalManagedThemeToCart` action. If you start loading products in `addExternalManagedThemeToCart`, it would be good to set `isLoadingCart` to `true` before that async load, so that the button is disabled immediately after clicking.

> Another solution would be having the `disabled` property set via state with a fixed initial state.

Yes, this also works, although it doesn't address the core issue. It's putting another hack on top of hacky code. You can do:
```js
const [ buttonDisabled, setButtonDisabled ] = useState( true );
const areProductsLoaded = ! isExternallyManagedTheme || Object.values( getProductsList( state ) ).length > 0;
useEffect( () => {
  if ( areProductsLoaded ) {
    setButtonDisabled( false );
  }
}, [ areProductsLoaded ] );
```
Here `buttonDisabled` always starts as `true`, and even if products are already loaded at initial render, it's reset to `false` only in an effect, a little bit later. Both server and client initial render have reliably always a disabled button.

The `disabled` prop for the button still needs to also take `isCartLoading` into account, so that it gets disabled also after clicking.

Reply:
I've used the second alternative adapted to a class component.

---

I did some tests with the first alternative, but after removing the ProductList check we would still need to add additional checks to keep the button disabled while its text is changing, and I was afraid to fall into the same case where we would have different initial states depending if its cached or not.

Reply:
> after removing the ProductList check we would still need to add additional checks to keep the button disabled while its text is changing

The text is changing as a reaction to the `/sites/:id/features` response arriving and being stored in the Redux state. Not as a reaction to the Products list. And nobody is waiting for the features.

It would be nice to implement the fixes that I suggested at some point. It would untangle some of the spaghetti that accumulated in the Themes code over time. It's not a blocker for this PR, but worth considering.